XiEditor/StyleMap.swift

```diff
@@ -108,4 +122,25 @@ class StyleMap {
                 applyStyle(string: string, id: styleSpan.style, range: styleSpan.range)
         }
     }
+
+    func stringWidth(str: NSString, style id: Int) -> CGFloat? {
+        if id >= map.count { return nil }
+        guard let style = map[id] else { return nil }
+
+        return str.size(withAttributes: style.attributes).width
+    }
+}
+
+func closestMatch(of font: NSFont, traits: NSFontTraitMask, weight: Int) -> NSFont? {
+    let fm = NSFontManager.shared()
+    var weight = weight
+    let fromWeight = fm.weight(of: font)
+    let direction = fromWeight > weight ? 1 : -1
+    while weight != fromWeight {
+        if let f = fm.font(withFamily: font.familyName ?? font.fontName, traits: traits, weight: weight, size: font.pointSize) {
+            return f
+        }
+        weight += direction
+    }
+    return nil
```

Comment:
This one looks for the font that best matches the targeted weight settings. (I should have added a doc comment before the function).